components/entities/api/src/main/java/org/phenotips/entities/PrimaryEntityConnectionsManager.java

```diff
@@ -0,0 +1,272 @@
+ * </p>
+ * <p>
+ * For the special case of *-to-one predicates, where each subject can be connected to at most one object, special
+ * simple methods are provided for convenience: {@link #get(PrimaryEntity)}, {@link #set(PrimaryEntity, PrimaryEntity)},
+ * and {@link #remove(PrimaryEntity)}. This type of predicate is called a <b>property</b>.
+ * </p>
+ * <p>
+ * Default implementations for many of the methods are provided, but should be re-implemented when a faster approach is
+ * possible.
+ * </p>
+ * <p>
+ * It is recommended that the name of each implementation follows the pattern: {@code subjectType-predicate-objectType},
+ * for example {@code family-contains-patient}, {@code patient-belongsTo-family}, {@code patient-uses-template}, and
+ * {@code template-usedBy-patient}.
+ * </p>
+ * <p>
+ * Each implementation should be thread-safe, but must not necessarily be atomic. This means that parallel calls to
+ * these methods shouldn't raise concurrency exceptions, but, for example, calling
+ * {@link #set(PrimaryEntity, PrimaryEntity)} for a subject that is connected to 4 objects before the call may end up
+ * breaking two of those connections, leaving the other two in place, and not creating the new connection, or if two
+ * calls to {@link #set(PrimaryEntity, PrimaryEntity)} occur at the same time, the subject may end up connected to
+ * either, none, or both of the new objects.
+ * </p>
+ *
+ * @param <S> the type of entities being the subject of the connection
+ * @param <O> the type of entities being the object of the connection
+ * @version $Id$
+ * @since 1.4
+ */
+@Unstable("New API introduced in 1.4")
+@Role
+public interface PrimaryEntityConnectionsManager<S extends PrimaryEntity, O extends PrimaryEntity>
+{
+    /**
+     * Creates a new connection between the Subject and the Object, if one didn't exist already.
+     *
+     * @param subject the subject to connect from
+     * @param object the object to connect to
+     * @return {@code true} if the connection was successfully created, or already existed, {@code false} if the
+     *         operation failed
+     */
+    boolean connect(@Nonnull S subject, @Nonnull O object);
+
+    /**
+     * Creates connections between the subject and each of the objects. Each individual connection creation is
+     * attempted, even if one of them fails.
+     *
+     * @param subject the subject to connect from
+     * @param objects a collection of objects to connect to
+     * @return {@code true} if all the connections were successfully created, {@code false} if any of the operations
+     *         failed
+     */
+    default boolean connectAll(@Nonnull S subject, @Nullable Collection<O> objects)
+    {
+        if (subject == null) {
+            throw new IllegalArgumentException();
+        }
+        return objects == null
+            || objects.stream().filter(Objects::nonNull).map(object -> connect(subject, object)).allMatch(b -> b);
```

Comment:
Not if we want to keep the behavior as described in the javadoc: "Each individual connection creation is attempted, even if one of them fails". `allMatch` says "May not evaluate the predicate on all elements if not necessary for determining the result".

Reply:
Hmm, then I think it will short-circuit on false regardless. So, instead you'd probably want ```objects.stream().filter(Objects::nonNull).map(object -> connect(subject, object)).reduce(Boolean.TRUE, (o1, o2) -> o1 && o2);```